feat(kcopy): new component [khcp-9464] (#1794)
Kong/kongponents: docs/components/copy.md

# Copy

> **KCopy** is a component that displays text content and copies it to clipboard.

Reviewer: suggestion: wording
```suggestion
**KCopy** is a component that displays provided text content and allows a user to copy it to their clipboard.
```

<KCopy :content="content" />

```html
<KCopy :content="content" />
```

## Props

### badgeText

Text displayed before the copyable content when `isBadge` is true.

<KCopy
  isBadge
  badge-text="Id"
  :content="content"
/>

```html
<KCopy
  isBadge
  badge-text="Id"
  :content="content"
/>
```

### content

The copyable text.

> ### contentTooltip

Reviewer: suggestion: can we group all of the `tooltip` props together?

Tooltip text displayed on hover over the `content`.

<KCopy
  truncate
  :content="content"
  content-tooltip="Hello!"
/>

```html
<KCopy
  truncate
  :content="content"
  content-tooltip="Hello!"
/>
```

### truncate

> Whether or not the content should be truncated.
> > Note: By default truncate is set to false.

Reviewer: ```suggestion
Whether or not the content should be truncated. Defaults to `false`.
```

<KCopy
  truncate
  :content="content"
/>

```html
<KCopy
  truncate
  :content="content"
/>
```

### truncationLimit

> Number of characters to truncate at.
> > Note: By default it is set to `8`.

Reviewer: ```suggestion
Number of characters to truncate at. Defaults to `8`.
```

<KCopy
  truncate
  :content="content"

>   truncation-limit=10

Reviewer: issue: You're missing quotes around the attribute value in a few places throughout the docs
```suggestion
  :truncation-limit="10"
```

/>

```html
<KCopy
  truncate
  :content="content"

>   truncation-limit=10

Reviewer: issue: since this is a number, you have to pass it with v-bind (also, you were missing quotes `"`)
```suggestion
  :truncation-limit="10"
```

> />

Reviewer: issue: This example sets the value to `10` but it's only showing `8` characters

Author: fixed.

```

### copyTooltip

> Tooltip text displayed on hover copy button. 
> > Note: If `badgeText` exists then text is `Copy <badge-text>` else `Copy`

Reviewer: ```suggestion
Tooltip text displayed on hover copy button. 

If the `badgeText` prop has a value, then the copy tooltip text is `Copy {badgeText}`; otherwise, `Copy`.
```

### successTooltip

> Tooltip text displayed on successful copy.
> > Note: By default `successTooltip` is set to `Copied!`.

Reviewer: ```suggestion
Tooltip text displayed on successful copy. Defaults to `Copied!`.
```

### format

Formatting for copyable content (default, hidden, redacted, deleted).

<KCopy
  format="hidden"
  :content="content"
/>

```html
<KCopy
  format="hidden"
  :content="content"
/>
```

<KCopy
  format="redacted"
  :content="content"
/>

```html
<KCopy
  format="redacted"
  :content="content"
/>
```

<KCopy
  format="deleted"
  :content="content"
/>

```html
<KCopy
  format="deleted"
  :content="content"
/>

> ```

Reviewer: docs: Can we add sub headers in this section for each type, along with an explanation?

### isBadge

> Whether or not to display as a badge.
> > Note: By default `isBadge` is set to false.

Reviewer: ```suggestion
Whether or not to display as a badge. Defaults to `false`.
```

<KCopy
  is-badge
  :content="content"
/>

```html
<KCopy
  is-badge

>   truncate

Reviewer: issue: you have `truncate` here but not in the actual usage on the line above

  :content="content"
/>
```

## Events

> - `@copied` -  Emitted when `copy button` is clicked; the payload is the copied text.

Reviewer: nit: can this event be `@copy` ?

Similar to native `@click` - not past-tense

# Copy

KCopy is a component that displays provided text and allows a user to copy it to their clipboard.

<KCopy :text="text" />

```html
<KCopy :text="text" />
```

## Props

### text

The copyable text.

### badge

Whether or not to display as a badge. Defaults to `false`.

<KCopy
  badge
  truncate
  :text="text"
/>

```html
<KCopy
  badge
  truncate
  :text="text"
/>
```

### badgeText

Text displayed before the copyable text when `badge` is true.

<KCopy
  badge
  badge-text="Id"
  :text="text"
/>

```html
<KCopy
  badge
  badge-text="Id"
  :text="text"
/>
```

### truncate

Whether or not the text should be truncated. Defaults to `false`.

<KCopy
  truncate
  :text="text"
/>

```html
<KCopy
  truncate
  :text="text"
/>
```

### truncationLimit

Number of characters to truncate at. Defaults to `8`.

<KCopy
  truncate
  :text="text"
  :truncation-limit="15"
/>

```html
<KCopy
  truncate
  :text="text"
  :truncation-limit="15"
/>
```

> ### copyTooltip

Reviewer: ![image](https://github.com/Kong/kongponents/assets/2229946/b02033cb-97ca-471b-83db-38eee2c159f6)
question: When I click to copy, the tooltip doesn't change when this prop is provided. Is that desired?

Author: Fixed.

Tooltip text displayed on hover copy button. 

> If the `badgeText` prop has a value, then the copy tooltip text is `Copy {badgeText}`; otherwise, `Copy`.

Reviewer: this doesn't mention that you're stripping certain characters from the `badgeText` for the tooltip?

Author: Added.
<img width="768" alt="image 1647" src="https://github.com/Kong/kongponents/assets/87779967/82c94888-9711-4cef-91d5-6300388e3f3c">

<KCopy
  :text="text"
  badge 
  badgeText="ID:" 
/>

```html
<KCopy
  :text="text"
  badge 
  badgeText="ID:" 
/>
```

### textTooltip

Tooltip text displayed on hover over the `text`.

<KCopy
  truncate
  :text="text"
  text-tooltip="Hello!"
/>

```html
<KCopy
  truncate
  :text="text"
  text-tooltip="Hello!"
/>
```

### successTooltip

> Tooltip text displayed on successful copy. Defaults to `Copied!`.

Reviewer: The success tooltip appears to be `Copied` by default (no exclamation mark)

Author: Added `!`

### monospace

An indicator of whether the copyable text has `JetBrains Mono` font or not. Defaults to `true`, but `false` if `badge` is `true`.

### format

Determines the display format of the copyable text. The component can take the following `format` values:

- `default`: displays regular uuid
- `hidden`: displays just a copy button without text
- `redacted`: displays `*****`
- `deleted`: displays `*<first-5-chars-of-uuid>`

<KCopy
  format="hidden"
  :text="text"
/>

```html
<KCopy
  format="hidden"
  :text="text"
/>
```

<KCopy
  format="redacted"
  :text="text"
/>

```html
<KCopy
  format="redacted"
  :text="text"
/>
```

<KCopy
  format="deleted"
  :text="text"
/>

```html
<KCopy
  format="deleted"
  :text="text"
/>

> ```

Reviewer: suggestion: Can you add a header over each example so you can tell which is which?

Author: Does this look ok? 
<img width="768" alt="image 1646" src="https://github.com/Kong/kongponents/assets/87779967/e43218fa-a14d-4335-b78b-f2838d0b1b3e">

Reviewer: nit: I'd just use h4/h5 tags, e.g.

`#### deleted` or h5 depending on where you are in the doc

Author: updated
<img width="1198" alt="image 1648" src="https://github.com/Kong/kongponents/assets/87779967/883e3b2a-d559-4ceb-a66a-a3045f600821">
